File: algorithms/04_java/numerical/NevilleInterpolation.java
// Neville's algorithm for polynomial interpolation.
public class NevilleInterpolation {
    static double neville(double[] xs, double[] ys, double x) {
        int n = xs.length;
        double[] table = ys.clone();
        for (int k = 1; k < n; k++)
            for (int i = 0; i < n - k; i++)
                table[i] = ((x - xs[i + k]) * table[i] + (xs[i] - x) * table[i + 1]) / (xs[i] - xs[i + k]);
        return table[0];
    }

    public static void main(String[] args) {
        assert Math.abs(neville(new double[] {0, 1, 2}, new double[] {1, 3, 2}, 1.5) - 2.875) < 1e-12;
        System.out.println("neville interpolation ok");
    }
}

File: algorithms/19_lua/numerical/neville_interpolation.lua
-- Neville's algorithm for polynomial interpolation.
local function neville_interpolation(xs, ys, x)
  local n = #xs
  local table_ = {}
  for i = 1, n do table_[i] = ys[i] end
  for k = 1, n - 1 do
    for i = 1, n - k do
      table_[i] = ((x - xs[i + k]) * table_[i] + (xs[i] - x) * table_[i + 1]) / (xs[i] - xs[i + k])
    end
  end
  return table_[1]
end

local value = neville_interpolation({ 0, 1, 2 }, { 1, 3, 2 }, 1.5)
assert(math.abs(value - 2.875) < 1e-12, "neville failed")
print("[Lua Neville Interpolation] verified")
